### DISCUSSION TITLE
Expanded Pattern Matching

### DISCUSSION THREAD

**Discussion by LupusInferni315** [dotnet/csharplang#6948] [Language Ideas]
With the current Pattern Matching the following is possible
```csharp
string value = /* Some Value */;
if(value is "Some Value" or "Some Other Value") {}
```

and I recently attempted to 'invert' the pattern and try this

```csharp
if(this or that is null) {}
```

and this

```csharp
if(null is this or that) {} // I understand why this one doesn't work, but it otherwise does fit the pattern.
```

which is not a valid pattern, I propose an addition to pattern matching such that either side of the pattern can be 'chained'.

```csharp
if(this and that is "") {}

if(this or that is "") {}
```

**Comment by DavidArno** [comment]
You seem to be proposing that:
```cs
x is null || y is null
```
could be expressed as:
```cs
x or y is null
```
Beyond that simple shortening of the syntax, what other use would such a pattern have? Could you give some other examples because - at the moment at least - it doesn't seem very compelling to me.

Indecently, I'd stay well clear of proposing `are` here. In English, if you have `subject1 or subject2` then if both subjects are singular, then the result remains singular, eg `the dog or cat is ...`, but if one is plural then the result is plural: `the dogs or cat are ...`. We really do not want the compiler having to try and work out whether variables are using singular or plural forms in order to make the syntax grammatically correct, especially as English is not the first language of a great many C# developers.

**Reply by LupusInferni315** [reply]
The main premise is indeed to shorten the syntax, the reason I had proposed the idea is because most patterns to a small degree read rather like English sentences, the proposal would have allowed more such patterns to be read as if they were simply that,

for example `if(this is null || that is null)` as a sentence would be written "if this is null or that is null" but `if(this or that is null)` doesn't require any additional 'transformation' to be read as a sentence.

Grammatically speaking the sentence `if [the] cat or dog are` the `are` doesn't refer to the number of cats or dogs but the number of 'arguments' being compared. The `are` part would only be used when placing multiple arguments before the comparison it would not change `if [the] cat is`.

Admittedly I had thought `are` would be required, and while for grammar's sake it would make some patterns make more sense it would ultimately not be required.

I have thus removed the `are` portion of the proposal.

**Comment by Joe4evr** [comment]
What you're asking for seems to be covered by placing the variables into a tuple and pattern matching over the tuple shape. [SharpLab.](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABATARgFgAoEgYgDsBXAGxoEMAjGiAAggqZdfoooHtg9YDH4UAziWwBmVnlYBhEgG8SrdXNmN+/GqwCS4gKIxgAC2gA5WjQAU/RgCsIAY2AB+VgA8krB87dPBABKNQ1VYg0ouQB2VlsfVhDWGHF46jpfAH1gvyh4rN8MmmCAbjD1AF8SCs1WbV0DY1MLKCMAWwAHYARbbHwABk9E/qGk0Mjw2qjsOITfZNT4gCJl7Nz+fNtC1lWy2uriQ5IgA==)

**Reply by LupusInferni315** [reply]
While it may allow the same idea to be done, it is much more complex and less readable that the current 'expected' solution, my idea was to expand the features of patterns to make reading them less complex and more easily understood at a glance.

**Comment by mungojam** [comment]
```cs
if(a and b is true) {}

if(a or b is false) {}
```

It would get confusing if your variables are booleans as somebody reading it might think that it is actually saying:

```cs
if(a and (b is true)) {}

if(a or (b is false)) {}
```

For the null `or` example, there is another way, but perhaps it's not as readable, depending on what a and b actually represent:

```cs
if(a ?? b is null) {}
```

**Reply by LupusInferni315** [reply]
For the purposes of this extended pattern matching, the idea was everything before is is the group to compare and everything after is the conditions.

```csharp
if(a or b is null)
```
would be 'if `either a or b` is null'.

```csharp
if(a and b is null)
```
would be 'if `both a and b` is null'.

```csharp
if((a or b) and c is null)
```
would be 'if `either a or b` is null and c is null'. just to include more complex cases.

However I had not considered that it could be interpreted more ways than this.
